File: 프로그래머스/level2/주차 요금 계산.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <map>

using namespace std;

int bTime, bFee, uTime, uFee;

int time(string in, string out) {
    int time = 0;
    time = (stoi(out.substr(0, 2)) - stoi(in.substr(0, 2))) * 60 + (stoi(out.substr(3, 2)) - stoi(in.substr(3, 2)));
    cout << "time : " << time << "\n";
    return time;
}

int fee(int t) {
    int num = bFee;
    if (bTime < t) {
        num += (t - bTime) / uTime * uFee;
        if ((t - bTime) % uTime != 0) {
            num += uFee;
        }
    }

    cout << "fee : " << num << "\n";

    return num;
}


vector<int> solution(vector<int> fees, vector<string> records) {
    vector<int> answer;
    bTime = fees[0]; bFee = fees[1]; uTime = fees[2]; uFee = fees[3];

    unordered_map<int, string> IN;
    map<int, int> mapTime;
    for (int i = 0; i < records.size(); i++) {
        stringstream ss(records[i]);
        string t, IO;
        int number;
        ss >> t >> number >> IO;
        if (IO == "IN") {
            IN[number] = t;
        }
        else {
            mapTime[number] += time(IN[number], t);
            IN[number] = "";
        }
    }

    for (auto m : IN) {
        if (m.second != "") {
            mapTime[m.first] += time(IN[m.first], "23:59");
        }
    }

    for (auto m : mapTime) {
        answer.push_back(fee(m.second));
    }


    return answer;
}

File: 프로그래머스(JAVA)/level2/주차 요금 계산.java
import java.util.*;

class Solution {
    public int bTime, bFee, uTime, uFee;
    
    public int time(String in, String out){
        int t = 0;
        String[] iSplit = in.split(":");
        String[] oSplit = out.split(":");
        t = (Integer.parseInt(oSplit[0])- Integer.parseInt(iSplit[0]))* 60 + (Integer.parseInt(oSplit[1])- Integer.parseInt(iSplit[1]));
        
        return t;
    }
    
    public int fee(int t){
        int f = bFee;
        
        if(bTime < t){
            f += ((t-bTime)/uTime)*uFee;
            if((t-bTime)%uTime!=0){
                f += uFee;
            }
        }
        System.out.println("fee : " + f);
        return f;
    }
    
    public int[] solution(int[] fees, String[] records) {
        int[] answer = {};
        bTime = fees[0]; bFee = fees[1]; uTime = fees[2]; uFee = fees[3];
        
        HashMap<String, String> In = new HashMap<>();
        Map<String, Integer> mapTime = new TreeMap<>();
        for(int i=0; i<records.length; i++){
            String[] ss = records[i].split(" ");
            if(ss[2].equals("IN")){
                In.put(ss[1], ss[0]);
            }
            else{
                mapTime.put(ss[1], mapTime.getOrDefault(ss[1], 0) + time(In.get(ss[1]), ss[0]));
                In.remove(ss[1]); //제거
            }
            
        }
        
        for(HashMap.Entry<String, String> e : In.entrySet()){
            mapTime.put(e.getKey(), mapTime.getOrDefault(e.getKey(), 0) + time(e.getValue(), "23:59"));
        }
        
        answer = new int[mapTime.size()];
        
        int i = 0;
        for(Map.Entry<String, Integer> e : mapTime.entrySet()){
            System.out.println(e.getValue());
            answer[i] = fee(e.getValue());
            i++;
        }
        
        return answer;
    }
    
}